File: src/test/java/net/i2p/crypto/eddsa/math/ed25519/Ed25519EdDSAFiniteFieldElementTest.java
/**
 * EdDSA-Java by str4d
 *
 * To the extent possible under law, the person who associated CC0 with
 * EdDSA-Java has waived all copyright and related or neighboring rights
 * to EdDSA-Java.
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <https://creativecommons.org/publicdomain/zero/1.0/>.
 *
 */
package net.i2p.crypto.eddsa.math.ed25519;

import net.i2p.crypto.eddsa.math.*;
import org.hamcrest.core.*;
import org.junit.*;

import java.math.BigInteger;

/**
 * Tests rely on the BigInteger class.
 */
public class Ed25519EdDSAFiniteFieldElementTest extends AbstractEdDSAFiniteFieldElementTest {

    protected FieldElement getRandomFieldElement() {
        return MathUtils.getRandomFieldElement();
    }

    protected BigInteger toBigInteger(final FieldElement f) {
        return MathUtils.toBigInteger(f);
    }

    protected BigInteger getQ() {
        return MathUtils.getQ();
    }

    protected EdDSAFiniteField getEdDSAFiniteField() {
        return MathUtils.getEdDSAFiniteField();
    }

    // region constructor

    @Test
    public void canConstructFieldElementFromArrayWithCorrectLength() {
        // Assert:
        new Ed25519FieldElement(MathUtils.getEdDSAFiniteField(), new int[10]);
    }

    @Test (expected = AssertionError.class)
    public void cannotConstructFieldElementFromArrayWithIncorrectLength() {
        // Assert:
        new Ed25519FieldElement(MathUtils.getEdDSAFiniteField(), new int[9]);
    }

    @Test (expected = AssertionError.class)
    public void cannotConstructFieldElementWithoutField() {
        // Assert:
        new Ed25519FieldElement(null, new int[9]);
    }

    // endregion

    // region isNonZero

    protected FieldElement getZeroFieldElement() {
        return new Ed25519FieldElement(MathUtils.getEdDSAFiniteField(), new int[10]);
    }

    protected FieldElement getNonZeroFieldElement() {
        final int[] t = new int[10];
        t[0] = 5;
        return new Ed25519FieldElement(MathUtils.getEdDSAFiniteField(), t);
    }

    // endregion

    // region toString

    @Test
    public void toStringReturnsCorrectRepresentation() {
        // Arrange:
        final byte[] bytes = new byte[32];
        for (int i = 0; 32 > i; i++) {
            bytes[i] = (byte)(i+1);
        }
        final FieldElement f = MathUtils.getEdDSAFiniteField().getEncoding().decode(bytes);

        // Act:
        final String fAsString = f.toString();
        final StringBuilder builder = new StringBuilder();
        builder.append("[Ed25519FieldElement val=");
        for (final byte b : bytes) {
            builder.append(String.format("%02x", Byte.valueOf(b)));
        }
        builder.append("]");

        // Assert:
        Assert.assertThat(fAsString, IsEqual.equalTo(builder.toString()));
    }

    // endregion
}

File: src/test/java/net/i2p/crypto/eddsa/math/ed25519/Ed25519EdDSAFiniteFieldElementTest.kt
/**
 * EdDSA-Java by str4d
 *
 * To the extent possible under law, the person who associated CC0 with
 * EdDSA-Java has waived all copyright and related or neighboring rights
 * to EdDSA-Java.
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <https:></https:>//creativecommons.org/publicdomain/zero/1.0/>.
 *
 */
package net.i2p.crypto.eddsa.math.ed25519

import net.i2p.crypto.eddsa.math.*
import org.hamcrest.core.*
import org.junit.*

import java.math.BigInteger

/**
 * Tests rely on the BigInteger class.
 */
class Ed25519EdDSAFiniteFieldElementTest : AbstractEdDSAFiniteFieldElementTest() {

    protected override val randomFieldElement: FieldElement
        get() = MathUtils.randomFieldElement

    protected override val q: BigInteger
        get() = MathUtils.q

    protected override val edDSAFiniteField: EdDSAFiniteField
        get() = MathUtils.edDSAFiniteField

    // endregion

    // region isNonZero

    protected override val zeroFieldElement: FieldElement
        get() = Ed25519FieldElement(MathUtils.edDSAFiniteField, IntArray(10))

    protected override val nonZeroFieldElement: FieldElement
        get() {
            val t = IntArray(10)
            t[0] = 5
            return Ed25519FieldElement(MathUtils.edDSAFiniteField, t)
        }

    override fun toBigInteger(f: FieldElement): BigInteger {
        return MathUtils.toBigInteger(f)
    }

    // region constructor

    @Test
    fun canConstructFieldElementFromArrayWithCorrectLength() {
        // Assert:
        Ed25519FieldElement(MathUtils.edDSAFiniteField, IntArray(10))
    }

    @Test(expected = AssertionError::class)
    fun cannotConstructFieldElementFromArrayWithIncorrectLength() {
        // Assert:
        Ed25519FieldElement(MathUtils.edDSAFiniteField, IntArray(9))
    }

    @Test(expected = AssertionError::class)
    fun cannotConstructFieldElementWithoutField() {
        // Assert:
        Ed25519FieldElement(null, IntArray(9))
    }

    // endregion

    // region toString

    @Test
    fun toStringReturnsCorrectRepresentation() {
        // Arrange:
        val bytes = ByteArray(32)
        var i = 0
        while (32 > i) {
            bytes[i] = (i + 1).toByte()
            i++
        }
        val f = MathUtils.edDSAFiniteField.encoding.decode(bytes)

        // Act:
        val fAsString = f.toString()
        val builder = StringBuilder()
        builder.append("[Ed25519FieldElement val=")
        for (b in bytes) {
            builder.append(String.format("%02x", java.lang.Byte.valueOf(b)))
        }
        builder.append("]")

        // Assert:
        Assert.assertThat(fAsString, IsEqual.equalTo(builder.toString()))
    }

    // endregion
}
